Support Mac OS

I don't have a mac to test on, need to find a helpful person.

I want to share the process I used on Mac OS X which did not succeed, but got further than some others have so-far indicated.

**Prerequisite** Ensure XCode is installed
```
$> brew intstall pkgconfig maven
$> git clone RichardWarburton/honest-profiler
$> git clone unittest-cpp/unittest-cpp
$> cd unittest-cpp/builds
$> cmake -G "Unix Makefiles" ../
$> cmake --build ./ --target install
(This gets the UnitTest++ libraries into /usr/local/include/UnitTest++)
$> cd ../../honest-profiler
$> export UNITTEST_INCLUDE_DIRS="/usr/local/include/UnitTest++"
$> export UNITTEST_LIBRARIES="UnitTest++"
$> cmake CMakeLists.txt
$> export export LC_ALL=C
$> make
```

Here I die like this:
```
$> make
Scanning dependencies of target lagent
[  5%] Building CXX object CMakeFiles/lagent.dir/src/main/cpp/agent.cpp.o
In file included from /Users/pepper/dev/honest-profiler/src/main/cpp/agent.cpp:8:
/Users/pepper/dev/honest-profiler/src/main/cpp/thread_map.h:28:1: error: thread-local storage is not supported for the current target
thread_local static map::GC::EpochType localEpoch = map::GC::kEpochInitial;
^
/Users/pepper/dev/honest-profiler/src/main/cpp/thread_map.h:29:1: error: thread-local storage is not supported for the current target
thread_local static int mapUsageCounter = 0;
^
2 errors generated.
make[2]: *** [CMakeFiles/lagent.dir/src/main/cpp/agent.cpp.o] Error 1
make[1]: *** [CMakeFiles/lagent.dir/all] Error 2
make: *** [all] Error 2
```

I'm not any kind of c++ developer, so I'm not sure what clang is trying to tell me, but it doesn't look happy.

--- a/src/main/cpp/agent.cpp
+++ b/src/main/cpp/agent.cpp
@@ -209,7 +209,6 @@ void JNICALL OnThreadStart(jvmtiEnv *jvmti_env, JNIEnv *jni_env, jthread thread)
 void JNICALL OnThreadEnd(jvmtiEnv *jvmti_env, JNIEnv *jni_env, jthread thread) {
     pthread_sigmask(SIG_BLOCK, &prof_signal_mask, NULL);
     threadMap.remove(jni_env);
-    threadMap.detach(); // detach must be called
 }
 
 static bool RegisterJvmti(jvmtiEnv *jvmti) {

--- a/src/main/cpp/agent.cpp
+++ b/src/main/cpp/agent.cpp
@@ -200,8 +200,7 @@ void JNICALL OnThreadStart(jvmtiEnv *jvmti_env, JNIEnv *jni_env, jthread thread)
                 }
             }
         }
-        // no need to call attach explicitly
-        threadMap.put(jni_env, thread);
+        threadMap.put(jni_env, thread_info.name);
     }
     pthread_sigmask(SIG_UNBLOCK, &prof_signal_mask, NULL);
 }

--- a/src/main/cpp/circular_queue.cpp
+++ b/src/main/cpp/circular_queue.cpp
@@ -2,20 +2,20 @@
 #include <iostream>
 #include <unistd.h>
 
-bool CircularQueue::push(const JVMPI_CallTrace &item) {
+bool CircularQueue::push(const JVMPI_CallTrace &item, ThreadBucket *info) {
     size_t currentInput;
     size_t nextInput;
     do {
-        currentInput = input.load();
+        currentInput = input.load(std::memory_order_relaxed);
         nextInput = advance(currentInput);
-        if (output.load() == nextInput) {
+        if (output.load(std::memory_order_relaxed) == nextInput) {
             return false;
         }
         // TODO: have someone review the memory ordering constraints
-    } while (!input.compare_exchange_strong(currentInput, nextInput,
-            std::memory_order_release,
-            std::memory_order_relaxed));
+    } while (!input.compare_exchange_strong(currentInput, nextInput, std::memory_order_relaxed));
     write(item, currentInput);
+    buffer[currentInput].info = info;
+    buffer[currentInput].is_committed.store(COMMITTED, std::memory_order_release);
 
     return true;
 }
@@ -33,23 +33,22 @@ void CircularQueue::write(const JVMPI_CallTrace &trace, const size_t slot) {
     buffer[slot].trace.frames = fb;
     buffer[slot].trace.num_frames = trace.num_frames;
     buffer[slot].trace.env_id = trace.env_id;
-    buffer[slot].is_committed.store(COMMITTED);
 }
 
 bool CircularQueue::pop() {
-    const auto current_output = output.load();
+    const auto current_output = output.load(std::memory_order_relaxed);
 
     // queue is empty
-    if (current_output == input.load()) {
+    if (current_output == input.load(std::memory_order_relaxed)) {
         return false;
     }
 
     // wait until we've finished writing to the buffer
-    while(buffer[current_output].is_committed.load() != COMMITTED) {
+    while (buffer[current_output].is_committed.load(std::memory_order_acquire) != COMMITTED) {
         usleep(1);
     }
 
-    listener_.record(buffer[current_output].trace);
+    listener_.record(buffer[current_output].trace, buffer[current_output].info);
     
     // 0 out all frames so the next write is clean
     JVMPI_CallFrame *fb = frame_buffer_[current_output];
@@ -57,11 +56,12 @@ bool CircularQueue::pop() {
     for (int frame_num = 0; frame_num < num_frames; ++frame_num) {
         memset(&(fb[frame_num]), 0, sizeof(JVMPI_CallFrame));
     }
+    buffer[current_output].info = nullptr;
 
     // ensure that the record is ready to be written to
-    buffer[current_output].is_committed.store(UNCOMMITTED);
+    buffer[current_output].is_committed.store(UNCOMMITTED, std::memory_order_release);
     // Signal that you've finished reading the record
-    output.store(advance(current_output));
+    output.store(advance(current_output), std::memory_order_relaxed);
 
     return true;
 }

--- a/src/main/cpp/circular_queue.h
+++ b/src/main/cpp/circular_queue.h
@@ -6,15 +6,9 @@
 #ifndef CIRCULAR_QUEUE_H
 #define CIRCULAR_QUEUE_H
 
+#include "thread_map.h"
 #include "stacktraces.h"
 #include <string.h>
-
-#if __GNUC__ == 4 && __GNUC_MINOR__ < 6 && !defined(__APPLE__) && !defined(__FreeBSD__) 
-  #include <cstdatomic>
-#else
-  #include <atomic>
-#endif
-
 #include <cstddef>
 
 const size_t Size = 1024;
@@ -27,7 +21,7 @@ const size_t Capacity = Size + 1;
 
 class QueueListener {
 public:
-    virtual void record(const JVMPI_CallTrace &item) = 0;
+    virtual void record(const JVMPI_CallTrace &item, ThreadBucket *info = nullptr) = 0;
 
     virtual ~QueueListener() {
     }
@@ -39,6 +33,7 @@ const int UNCOMMITTED = 0;
 struct TraceHolder {
     std::atomic<int> is_committed;
     JVMPI_CallTrace trace;
+    ThreadBucket *info;
 };
 
 class CircularQueue {
@@ -55,7 +50,7 @@ public:
             delete[] frame_buffer_[i];
     }
 
-    bool push(const JVMPI_CallTrace &item);
+    bool push(const JVMPI_CallTrace &item, ThreadBucket *info = nullptr);
 
     bool pop();
 

--- a/src/main/cpp/concurrent_map.h
+++ b/src/main/cpp/concurrent_map.h
@@ -252,12 +252,12 @@ public:
 		}
 	}
 
-	static InsertOutcome insertOrUpdate(HashTable *root, KeyType key, ValueType value, HashFunction hasher) {
+	static InsertOutcome insertOrUpdate(HashTable *root, KeyType key, ValueType value, HashFunction hasher, ValueType& oldValue) {
 		HashType hash = hasher(key);
-		return insertOrUpdate(root, hash, value);
+		return insertOrUpdate(root, hash, value, oldValue);
 	}
 
-	static InsertOutcome insertOrUpdate(HashTable *root, HashType tHash, ValueType value) {
+	static InsertOutcome insertOrUpdate(HashTable *root, HashType tHash, ValueType value, ValueType& oldValue) {
 		int i = tHash & root->sizeMask;
 		int jumps = 0;
 		DeltaType delta = 0;
@@ -276,7 +276,7 @@ public:
 
 				// all items in chain are already allocated
 				if (bucketHash == tHash) { // allocated bucket found
-					ValueType oldValue = prev->value.load(std::memory_order_relaxed);
+					oldValue = prev->value.load(std::memory_order_relaxed);
 					if (oldValue == MapValMove) {
 						return INSERT_HELP_MIGRATION; // indicate overflow
 					} else if (prev->value.compare_exchange_strong(oldValue, value, std::memory_order_acq_rel)) {
@@ -330,7 +330,7 @@ public:
 
 			if (bucketHash == tHash) {
 				// by chance bucket was allocated but unnoticed or allocated by code above
-				ValueType oldValue = entr->value.load(std::memory_order_relaxed); // we are only interested in address
+				oldValue = entr->value.load(std::memory_order_relaxed); // we are only interested in address
 				if (oldValue == MapValMove) {
 					return INSERT_HELP_MIGRATION; // indicate overflow
 				} else if (entr->value.compare_exchange_strong(oldValue, value, std::memory_order_acq_rel)) {
@@ -497,7 +497,8 @@ end_migration:
 
 					// only 1 thread can migrate bucket at time, so srcValue != MapValMove and srcHash is not in dest
 					if (srcValue != MapValEmpty) {
-						InsertOutcome res = LockFreeMapPrimitives::insertOrUpdate(dest, srcHash, srcValue);
+						ValueType oldValue;
+						InsertOutcome res = LockFreeMapPrimitives::insertOrUpdate(dest, srcHash, srcValue, oldValue);
 						if (res == INSERT_OVERFLOW) {
 							entry->value.store(srcValue, std::memory_order_release); // return replaced value
 							return true; // overflow
@@ -516,7 +517,7 @@ end_migration:
 
 class AbstractMapProvider {
 public:
-	virtual void put(KeyType key, ValueType value) = 0;
+	virtual ValueType put(KeyType key, ValueType value) = 0;
 	virtual ValueType get(KeyType key) = 0;
 	virtual ValueType remove(KeyType key) = 0;
 	virtual ~AbstractMapProvider() {}
@@ -583,14 +584,15 @@ public:
 		delete current.load(std::memory_order_consume);
 	}
 
-	void put(KeyType key, ValueType value) {
+	ValueType put(KeyType key, ValueType value) {
 		bool doubleSize = false;
 		while (true) {
 			HashTable *root = current.load(std::memory_order_consume);
-			InsertOutcome res = LockFreeMapPrimitives::insertOrUpdate(root, key, value, hasher);
+			ValueType oldValue;
+			InsertOutcome res = LockFreeMapPrimitives::insertOrUpdate(root, key, value, hasher, oldValue);
 
 			if (res == INSERT_OK) {
-				return;
+				return oldValue;
 			} else if (res == INSERT_OVERFLOW) {
 				migrationStart(root, doubleSize);
 			}

--- a/src/main/cpp/concurrent_map.h
+++ b/src/main/cpp/concurrent_map.h
@@ -9,9 +9,9 @@
  */
 
 #if __GNUC__ == 4 && __GNUC_MINOR__ < 6 && !defined(__APPLE__) && !defined(__FreeBSD__)
-#include <cstdatomic>
+#	include <cstdatomic>
 #else
-#include <atomic>
+#	include <atomic>
 #endif
 
 #include <vector>
@@ -147,8 +147,8 @@ class GC {
 private:
 	std::mutex mutex;
 	int totalThreads;
-	int globalEpoch;
-	int remaining;
+	std::atomic_int globalEpoch;
+	std::atomic_int remaining;
 	std::vector<HashTable*> recentGarbage;
 	std::vector<HashTable*> oldGarbage;
 #ifdef DEBUG_MAP_GC
@@ -165,8 +165,8 @@ public:
 			delete oldGarbage.back();
 			oldGarbage.pop_back();
 		}
-		globalEpoch++;
-		remaining = totalThreads;
+		globalEpoch.fetch_add(1, std::memory_order_relaxed);
+		remaining.store(totalThreads, std::memory_order_relaxed);
 		oldGarbage = std::move(recentGarbage);
 		recentGarbage.clear();
 	}
@@ -180,28 +180,44 @@ public:
 	EpochType attachThread() {
 		std::lock_guard<std::mutex> guard(mutex);
 		totalThreads++;
-		remaining++;
-		return globalEpoch - 1; // force thread to move to global epoch at least once
+		remaining.fetch_add(1, std::memory_order_relaxed);
+		return globalEpoch.load(std::memory_order_relaxed) - 1; // force thread to move to global epoch at least once
 	}
 
 	void detachThread(EpochType &localEpoch) {
 		std::lock_guard<std::mutex> guard(mutex);
 		totalThreads--;
+		int rem;
 		if (localEpoch < globalEpoch) {
-			remaining--;
-			if (remaining == 0)
-				newEpoch();
+			rem = remaining.fetch_sub(1, std::memory_order_relaxed) - 1;
+		} else {
+			rem = remaining.load(std::memory_order_relaxed);
 		}
+		if (rem == 0) 
+			newEpoch();
 		localEpoch = kEpochInitial;
 	}
 
 	void safepoint(EpochType &localEpoch) {
 		std::lock_guard<std::mutex> guard(mutex);
+		int rem;
+
 		if (localEpoch < globalEpoch) {
 			localEpoch = globalEpoch;
-			remaining--;
-			if (remaining == 0)
-				newEpoch();
+			rem = remaining.fetch_sub(1, std::memory_order_relaxed) - 1;
+		} else {
+			rem = remaining.load(std::memory_order_relaxed);
+		}
+		if (rem == 0)
+			newEpoch();
+	}
+
+	/* for reads within signal handler */
+	void ss_safepoint(EpochType &localEpoch) {
+		EpochType global = globalEpoch.load(std::memory_order_relaxed);
+		if (localEpoch < global) { // no thread can move global epoch since remaining > 0
+			remaining.fetch_sub(1, std::memory_order_relaxed);
+			localEpoch = global;
 		}
 	}
 

--- a/src/main/cpp/log_writer.cpp
+++ b/src/main/cpp/log_writer.cpp
@@ -22,29 +22,6 @@ void LogWriter::writeValue(const T &value) {
     }
 }
 
-static int64_t getThreadId(JNIEnv *env_id, jvmtiEnv *tiEnv, ThreadMap &tMap) {
-    ThreadBucket *tInfo = tMap.get(env_id);
-    if (tInfo && tInfo->tid > 0) {
-        return (int64_t)tInfo->tid;
-    }
-    return (int64_t)env_id;
-}
-
-static void getThreadName(JNIEnv *env_id, jvmtiEnv *tiEnv, ThreadMap &tMap, char *buff, int buffsz) {
-    jvmtiThreadInfo thread_info;
-    char *src = (char*)"Unknown";
-    ThreadBucket *tInfo = tMap.get(env_id);
-    if (tInfo && tInfo->tid > 0) {
-        int error = tiEnv->GetThreadInfo(tInfo->thread, &thread_info);
-        if (error == JNI_OK) {
-            src = thread_info.name;
-        }
-    }
-    int symbols = std::min((int)strlen(src), buffsz - 1);
-    strncpy(buff, src, symbols);
-    buff[symbols] = '\0';
-}
-
 static jint bci2line(jint bci, jvmtiLineNumberEntry *table, jint entry_count) {
 	jint line_number = -101;
 	if ( entry_count == 0 ) {
@@ -101,12 +78,17 @@ jint LogWriter::getLineNo(jint bci, jmethodID methodId) {
     return lineno;
 }
 
-void LogWriter::record(const JVMPI_CallTrace &trace) {
-    constexpr const int size = 128;
-    char namebuff[size];
+void LogWriter::record(const JVMPI_CallTrace &trace, ThreadBucket *info) {
+    map::HashType threadId = (map::HashType) trace.env_id;
+    // char *src = (char*)"";
 
-    int64_t threadId = getThreadId(trace.env_id, jvmti_, tMap_);
-    getThreadName(trace.env_id, jvmti_, tMap_, namebuff, size);
+    if (info) {
+        threadId = (map::HashType) info->tid;
+        // src = info->name;
+    }
+
+    // std::cout << "#### thread: " << trace.env_id << ", tid: " << threadId << ", src: " << src << std::endl;
+
     recordTraceStart(trace.num_frames, threadId);
 
     for (int i = 0; i < trace.num_frames; i++) {
@@ -124,14 +106,9 @@ void LogWriter::record(const JVMPI_CallTrace &trace) {
         }
         inspectMethod(methodId, frame);
     }
-}
 
-void LogWriter::startup() {
-    tMap_.attachReader();
-}
-
-void LogWriter::terminate() {
-    tMap_.detachReader();
+    if (info != nullptr)
+        info->release();
 }
 
 void LogWriter::inspectMethod(const method_id methodId,
@@ -144,7 +121,7 @@ void LogWriter::inspectMethod(const method_id methodId,
     frameLookup_(frame, jvmti_, *this);
 }
 
-void LogWriter::recordTraceStart(const jint numFrames, const int64_t threadId) {
+void LogWriter::recordTraceStart(const jint numFrames, const map::HashType threadId) {
     output_.put(TRACE_START);
     writeValue(numFrames);
     writeValue(threadId);
@@ -173,7 +150,7 @@ void LogWriter::writeWithSize(const char *value) {
     output_.write(value, size);
 }
 
-void LogWriter::recordNewMethod(const int64_t methodId, const char *fileName,
+void LogWriter::recordNewMethod(const map::HashType methodId, const char *fileName,
         const char *className, const char *methodName) {
     output_.put(NEW_METHOD);
     writeValue(methodId);

--- a/src/main/cpp/log_writer.cpp
+++ b/src/main/cpp/log_writer.cpp
@@ -126,8 +126,12 @@ void LogWriter::record(const JVMPI_CallTrace &trace) {
     }
 }
 
+void LogWriter::startup() {
+    tMap_.attachReader();
+}
+
 void LogWriter::terminate() {
-    tMap_.detach();
+    tMap_.detachReader();
 }
 
 void LogWriter::inspectMethod(const method_id methodId,

--- a/src/main/cpp/log_writer.h
+++ b/src/main/cpp/log_writer.h
@@ -51,6 +51,8 @@ public:
 
     virtual void record(const JVMPI_CallTrace &trace);
 
+    virtual void startup();    
+
     virtual void terminate();
 
     void recordTraceStart(const jint num_frames, const int64_t threadId);

--- a/src/main/cpp/log_writer.h
+++ b/src/main/cpp/log_writer.h
@@ -45,8 +45,8 @@ class LogWriter : public QueueListener, public MethodListener {
 
 public:
     explicit LogWriter(ostream &output, GetFrameInformation frameLookup,
-            jvmtiEnv *jvmti, ThreadMap &tMap)
-            : output_(output), frameLookup_(frameLookup), jvmti_(jvmti), tMap_(tMap) {
+            jvmtiEnv *jvmti)
+            : output_(output), frameLookup_(frameLookup), jvmti_(jvmti) {
     }
 
     virtual void record(const JVMPI_CallTrace &trace, ThreadBucket *info = nullptr);
@@ -68,8 +68,6 @@ private:
     GetFrameInformation frameLookup_;
 
     jvmtiEnv *jvmti_;
-
-    ThreadMap &tMap_;
 
     unordered_set<method_id> knownMethods;
 

--- a/src/main/cpp/log_writer.h
+++ b/src/main/cpp/log_writer.h
@@ -49,13 +49,9 @@ public:
             : output_(output), frameLookup_(frameLookup), jvmti_(jvmti), tMap_(tMap) {
     }
 
-    virtual void record(const JVMPI_CallTrace &trace);
+    virtual void record(const JVMPI_CallTrace &trace, ThreadBucket *info = nullptr);
 
-    virtual void startup();    
-
-    virtual void terminate();
-
-    void recordTraceStart(const jint num_frames, const int64_t threadId);
+    void recordTraceStart(const jint num_frames, const map::HashType threadId);
 
     // method are unique pointers, use a long to standardise
     // between 32 and 64 bits

--- a/src/main/cpp/processor.cpp
+++ b/src/main/cpp/processor.cpp
@@ -28,7 +28,6 @@ TRACE_DEFINE_END(Processor, kTraceProcessorTotal);
 
 void Processor::run() {
     int popped = 0;
-    logWriter_.startup();
 
     while (true) {
         while (buffer_.pop()) {
@@ -50,7 +49,6 @@ void Processor::run() {
     }
 
     handler_.stopSigprof();
-    logWriter_.terminate(); // make sure processing thread is detached from thread map
     workerDone.clear(std::memory_order_relaxed);
     // no shared data access after this point, can be safely deleted
 }

--- a/src/main/cpp/processor.cpp
+++ b/src/main/cpp/processor.cpp
@@ -28,6 +28,7 @@ TRACE_DEFINE_END(Processor, kTraceProcessorTotal);
 
 void Processor::run() {
     int popped = 0;
+    logWriter_.startup();
 
     while (true) {
         while (buffer_.pop()) {

--- a/src/main/cpp/profiler.cpp
+++ b/src/main/cpp/profiler.cpp
@@ -72,6 +72,7 @@ void Profiler::handle(int signum, siginfo_t *info, void *context) {
     IMPLICITLY_USE(signum);
     IMPLICITLY_USE(info);
     SimpleSpinLockGuard<false> guard(ongoingConf); // sync buffer
+    ThreadBucket *threadInfo;
 
     // sample data structure
     STATIC_ARRAY(frames, JVMPI_CallFrame, configuration_->maxFramesToCapture, MAX_FRAMES_TO_CAPTURE);
@@ -81,13 +82,15 @@ void Profiler::handle(int signum, siginfo_t *info, void *context) {
     JNIEnv *jniEnv = getJNIEnv(jvm_);
     if (jniEnv == NULL) {
         trace.num_frames = -3; // ticks_unknown_not_Java
+        threadInfo = nullptr;
     } else {
         trace.env_id = jniEnv;
         ASGCTType asgct = Asgct::GetAsgct();
         (*asgct)(&trace, configuration_->maxFramesToCapture, context);
+        threadInfo = tMap_.get(jniEnv);
     }
     // log all samples, failures included, let the post processing sift through the data
-    buffer->push(trace);
+    buffer->push(trace, threadInfo);
 }
 
 bool Profiler::start(JNIEnv *jniEnv) {

--- a/src/main/cpp/profiler.cpp
+++ b/src/main/cpp/profiler.cpp
@@ -256,7 +256,7 @@ void Profiler::configure() {
             // The JVM will still continue to run though; could call abort() to terminate the JVM abnormally.
             logError("ERROR: Failed to open file %s for writing\n", fileName);
         }
-        writer = new LogWriter(*logFile, &Profiler::lookupFrameInformation, jvmti_, tMap_);
+        writer = new LogWriter(*logFile, &Profiler::lookupFrameInformation, jvmti_);
     }
 
     needsUpdate = needsUpdate || configuration_->maxFramesToCapture != liveConfiguration->maxFramesToCapture;

--- a/src/main/cpp/thread_map.h
+++ b/src/main/cpp/thread_map.h
@@ -3,30 +3,38 @@
 
 #include <jvmti.h>
 #include <jni.h>
+#include <string.h>
 #include "concurrent_map.h"
 
 
 int gettid();
 
+
 template <typename PType>
 struct PointerHasher {
-	/* pointer hash bijection (collision-free) */
+	/* Numerical Recipes, 3rd Edition */
 	static int64_t hash(void *p) {
-		// remove LSB zeros in pointers
-		return ((int64_t)p) / sizeof(PType);
+		int64_t v = (int64_t)p / sizeof(PType);
+		v = v * 3935559000370003845 + 2691343689449507681;
+
+  		v ^= v >> 21;
+  		v ^= v << 37;
+  		v ^= v >>  4;
+
+  		v *= 4768777513237032717;
+
+  		v ^= v << 20;
+  		v ^= v >> 41;
+  		v ^= v <<  5;
+
+  		return v;
 	}
 };
 
 const int kInitialMapSize = 256;
-const int KSafepointTrigger = 32; // each 32nd read results in safepoint
 
 class GCHelper {
 public:
-	static void safepointOnDemand(map::GC::EpochType &localEpoch, unsigned int &count) {
-		if (!(count++ & (KSafepointTrigger - 1))) // count = 0 (mod KSafepointTrigger)
-			map::DefaultGC.safepoint(localEpoch);
-	}
-
 	static map::GC::EpochType attach() {
 		return map::DefaultGC.attachThread();
 	}
@@ -39,25 +47,40 @@ public:
 	static void safepoint(map::GC::EpochType &localEpoch) {
 		map::DefaultGC.safepoint(localEpoch);
 	}
+
+	static void signalSafepoint(map::GC::EpochType &localEpoch) {
+		map::DefaultGC.ss_safepoint(localEpoch);
+	}
 };
 
 struct ThreadBucket {
 	const int tid;
-	const jthread thread;
-	map::GC::EpochType localEpoch;
+	char *name;
 	std::atomic_int refs;
+	map::GC::EpochType localEpoch;
 
-	ThreadBucket(int id, jthread thr) : tid(id), thread(thr), localEpoch(GCHelper::attach()), refs(1) {
+	ThreadBucket(int id, const char *n) : tid(id), refs(1), localEpoch(GCHelper::attach()) {
+		int len = strlen(n) + 1;
+		name = new char[len];
+		std::copy(n, n + len, name);
+	}
+
+	void release() {
+		int prev = refs.fetch_sub(1, std::memory_order_acquire);
+		if (prev == 1)
+			delete this;
+	}
+
+	~ThreadBucket() {
+		delete[] name;
 	}
 };
 
-// MWSR
+
 template <typename MapProvider>
 class ThreadMapBase {
 private:
 	MapProvider map;
-	map::GC::EpochType readerEpoch;
-	unsigned int readerCounter;
 
 	static ThreadBucket *acq_bucket(ThreadBucket *tb) {
 		if (tb != nullptr) {
@@ -70,62 +93,39 @@ private:
 		return nullptr;
 	}
 
-	static void rel_bucket(ThreadBucket *tb, JNIEnv *jni_env) {
-		if (tb != nullptr) {
-			int prev = tb->refs.fetch_sub(1, std::memory_order_acquire);
-			if (prev == 1) {
-				jni_env->DeleteGlobalRef(tb->thread);
-				delete tb;
-			}
-		}
-	}
-
 public:
 
-	ThreadMapBase(int capacity = kInitialMapSize) : map(capacity) {
-		readerEpoch = map::GC::kEpochInitial;
-		readerCounter = 0;
+	ThreadMapBase(int capacity = kInitialMapSize) : map(capacity) {}
+
+	void put(JNIEnv *jni_env, const char *name) {
+		put(jni_env, name, gettid());
 	}
 
-	void put(JNIEnv *jni_env, jthread thread) {
-		put(jni_env, thread, gettid());
-	}
-
-	void put(JNIEnv *jni_env, jthread thread, int tid, bool globalRef = true) {
+	void put(JNIEnv *jni_env, const char *name, int tid) {
 		// constructor == call to acquire
-		ThreadBucket *info = new ThreadBucket(tid, globalRef ? thread : jni_env->NewGlobalRef(thread));
+		ThreadBucket *info = new ThreadBucket(tid, name);
 		ThreadBucket *old = (ThreadBucket*)map.put((map::KeyType)jni_env, (map::ValueType)info);
-		rel_bucket(old, jni_env);
+		if (old != nullptr)
+			old->release();
 		GCHelper::safepoint(info->localEpoch); // each thread inserts once
 	}
 
 	ThreadBucket *get(JNIEnv *jni_env) {
-		GCHelper::safepointOnDemand(readerEpoch, readerCounter);
-		return acq_bucket((ThreadBucket*)map.get((map::KeyType)jni_env));
+		ThreadBucket *info = acq_bucket((ThreadBucket*)map.get((map::KeyType)jni_env));
+		if (info != nullptr)
+			GCHelper::signalSafepoint(info->localEpoch);
+		return info;
 	}
 
 	void remove(JNIEnv *jni_env) {
 		ThreadBucket *info = (ThreadBucket*)map.remove((map::KeyType)jni_env);
-		if (info) {
+		if (info != nullptr) {
 			GCHelper::detach(info->localEpoch);
-			rel_bucket(info, jni_env);
-		}
-	}
-
-	void attachReader() {
-		if (readerEpoch == map::GC::kEpochInitial) {
-			readerEpoch = GCHelper::attach();
-			readerCounter = 0;
-		}
-	}
-
-	void detachReader() {
-		if (readerEpoch != map::GC::kEpochInitial) {
-			GCHelper::detach(readerEpoch);
+			info->release();
 		}
 	}
 };
 
-typedef ThreadMapBase<map::ConcurrentMapProvider<PointerHasher<JNIEnv>, false> > ThreadMap;
+typedef ThreadMapBase<map::ConcurrentMapProvider<PointerHasher<JNIEnv>, true> > ThreadMap;
 
 #endif

--- a/src/main/cpp/thread_map.h
+++ b/src/main/cpp/thread_map.h
@@ -8,11 +8,6 @@
 
 int gettid();
 
-struct ThreadBucket {
-	int tid;
-	jthread thread;
-};
-
 template <typename PType>
 struct PointerHasher {
 	/* pointer hash bijection (collision-free) */
@@ -25,76 +20,109 @@ struct PointerHasher {
 const int kInitialMapSize = 256;
 const int KSafepointTrigger = 32; // each 32nd read results in safepoint
 
-thread_local static map::GC::EpochType localEpoch = map::GC::kEpochInitial;
-thread_local static int mapUsageCounter = 0;
-
 class GCHelper {
 public:
-	static void safepointOnDemand() {
-		if (!(mapUsageCounter++ & (KSafepointTrigger - 1))) // mapUsageCounter = 0 (mod KSafepointTrigger)
+	static void safepointOnDemand(map::GC::EpochType &localEpoch, unsigned int &count) {
+		if (!(count++ & (KSafepointTrigger - 1))) // count = 0 (mod KSafepointTrigger)
 			map::DefaultGC.safepoint(localEpoch);
 	}
 
-	static void attach() {
-		if (localEpoch == map::GC::kEpochInitial)
-			localEpoch = map::DefaultGC.attachThread();
+	static map::GC::EpochType attach() {
+		return map::DefaultGC.attachThread();
 	}
 
-	static void detach() {
+	static void detach(map::GC::EpochType &localEpoch) {
 		if (localEpoch != map::GC::kEpochInitial)
 			map::DefaultGC.detachThread(localEpoch);
 	}
 
-	static void safepoint() {
-		if (localEpoch == map::GC::kEpochInitial) attach();
+	static void safepoint(map::GC::EpochType &localEpoch) {
 		map::DefaultGC.safepoint(localEpoch);
 	}
 };
 
+struct ThreadBucket {
+	const int tid;
+	const jthread thread;
+	map::GC::EpochType localEpoch;
+	std::atomic_int refs;
+
+	ThreadBucket(int id, jthread thr) : tid(id), thread(thr), localEpoch(GCHelper::attach()), refs(1) {
+	}
+};
+
+// MWSR
 template <typename MapProvider>
 class ThreadMapBase {
 private:
 	MapProvider map;
+	map::GC::EpochType readerEpoch;
+	unsigned int readerCounter;
+
+	static ThreadBucket *acq_bucket(ThreadBucket *tb) {
+		if (tb != nullptr) {
+			int prev = tb->refs.fetch_add(1, std::memory_order_relaxed);
+			if (prev > 0) {
+				return tb;
+			}
+			tb->refs.fetch_sub(1, std::memory_order_relaxed);
+		}
+		return nullptr;
+	}
+
+	static void rel_bucket(ThreadBucket *tb, JNIEnv *jni_env) {
+		if (tb != nullptr) {
+			int prev = tb->refs.fetch_sub(1, std::memory_order_acquire);
+			if (prev == 1) {
+				jni_env->DeleteGlobalRef(tb->thread);
+				delete tb;
+			}
+		}
+	}
 
 public:
 
-	ThreadMapBase(int capacity = kInitialMapSize) : map(capacity) {}
+	ThreadMapBase(int capacity = kInitialMapSize) : map(capacity) {
+		readerEpoch = map::GC::kEpochInitial;
+		readerCounter = 0;
+	}
 
 	void put(JNIEnv *jni_env, jthread thread) {
 		put(jni_env, thread, gettid());
 	}
 
 	void put(JNIEnv *jni_env, jthread thread, int tid, bool globalRef = true) {
-		GCHelper::attach();
-		ThreadBucket *info = new ThreadBucket;
-		info->tid = tid;
-		info->thread = globalRef ? thread : jni_env->NewGlobalRef(thread);
-		map.put((map::KeyType)jni_env, (map::ValueType)info);
-		GCHelper::safepoint(); // each thread inserts once
+		// constructor == call to acquire
+		ThreadBucket *info = new ThreadBucket(tid, globalRef ? thread : jni_env->NewGlobalRef(thread));
+		ThreadBucket *old = (ThreadBucket*)map.put((map::KeyType)jni_env, (map::ValueType)info);
+		rel_bucket(old, jni_env);
+		GCHelper::safepoint(info->localEpoch); // each thread inserts once
 	}
 
 	ThreadBucket *get(JNIEnv *jni_env) {
-		GCHelper::attach();
-		GCHelper::safepointOnDemand();
-		return reinterpret_cast<ThreadBucket*>(map.get((map::KeyType)jni_env));
+		GCHelper::safepointOnDemand(readerEpoch, readerCounter);
+		return acq_bucket((ThreadBucket*)map.get((map::KeyType)jni_env));
 	}
 
 	void remove(JNIEnv *jni_env) {
-		GCHelper::attach();
 		ThreadBucket *info = (ThreadBucket*)map.remove((map::KeyType)jni_env);
 		if (info) {
-			jni_env->DeleteGlobalRef(info->thread);
-			delete info;
+			GCHelper::detach(info->localEpoch);
+			rel_bucket(info, jni_env);
 		}
-		GCHelper::safepoint(); // each thread deletes once
 	}
 
-	void attach() {
-		GCHelper::attach();
+	void attachReader() {
+		if (readerEpoch == map::GC::kEpochInitial) {
+			readerEpoch = GCHelper::attach();
+			readerCounter = 0;
+		}
 	}
 
-	void detach() {
-		GCHelper::detach();
+	void detachReader() {
+		if (readerEpoch != map::GC::kEpochInitial) {
+			GCHelper::detach(readerEpoch);
+		}
 	}
 };
 

--- a/src/test/cpp/fixtures.h
+++ b/src/test/cpp/fixtures.h
@@ -8,7 +8,7 @@ class ItemHolder : public QueueListener {
 public:
   explicit ItemHolder() {}
 
-  virtual void record(const JVMPI_CallTrace &trace) {
+  virtual void record(const JVMPI_CallTrace &trace, ThreadBucket *info) {
     CHECK_EQUAL(2, trace.num_frames);
     CHECK_EQUAL((JNIEnv *)envId, trace.env_id);
 

--- a/src/test/cpp/test_log_writer.cpp
+++ b/src/test/cpp/test_log_writer.cpp
@@ -25,8 +25,7 @@ bool stubFrameInformation(const JVMPI_CallFrame &frame, jvmtiEnv *jvmti,
   char buffer[100] = {};                                                       \
   ostreambuf<char> outputBuffer(buffer, sizeof(buffer));                       \
   ostream output(&outputBuffer);                                               \
-  ThreadMap emptyMapStub;                                                      \
-  LogWriter logWriter(output, &stubFrameInformation, NULL, emptyMapStub);      \
+  LogWriter logWriter(output, &stubFrameInformation, NULL);                    \
   CircularQueue *queue = new CircularQueue(logWriter, DEFAULT_MAX_FRAMES_TO_CAPTURE);
 
 #define done() delete queue;
@@ -157,9 +156,8 @@ bool dumpStubFrameInformation(const JVMPI_CallFrame &frame, jvmtiEnv *jvmti,
 TEST(DumpTestFile) {
   givenStackTrace();
 
-  ThreadMap tMap;
   ofstream output("dump.hpl", ofstream::out | ofstream::binary);
-  LogWriter logWriter(output, &dumpStubFrameInformation, NULL, tMap);
+  LogWriter logWriter(output, &dumpStubFrameInformation, NULL);
 
   logWriter.record(trace);
   logWriter.record(trace);

--- a/src/test/cpp/test_log_writer.cpp
+++ b/src/test/cpp/test_log_writer.cpp
@@ -29,7 +29,7 @@ bool stubFrameInformation(const JVMPI_CallFrame &frame, jvmtiEnv *jvmti,
   LogWriter logWriter(output, &stubFrameInformation, NULL, emptyMapStub);      \
   CircularQueue *queue = new CircularQueue(logWriter, DEFAULT_MAX_FRAMES_TO_CAPTURE);
 
-#define done() logWriter.terminate(); delete queue;
+#define done() delete queue;
 
 TEST(RecordsStartOfStackTrace) {
   givenLogWriter();
@@ -163,5 +163,4 @@ TEST(DumpTestFile) {
 
   logWriter.record(trace);
   logWriter.record(trace);
-  logWriter.terminate(); // detach a thread from map
 }

--- a/src/test/cpp/test_maps.cpp
+++ b/src/test/cpp/test_maps.cpp
@@ -33,32 +33,32 @@ void deallocateTestBuffer(void **b, size_t size) {
 #define IDX(rev, i, sz) ((rev) ? (sz - 1 - i) : (i))
 
 void mapWriter(AbstractMapProvider &map, void **keys, void **values, std::atomic<void*> *result, size_t size, bool reverse=false) {
-	GCHelper::attach();
+	map::GC::EpochType id = GCHelper::attach();
 	for (int i = 0; i < size; i++) {
 		map.put(keys[IDX(reverse, i, size)], values[IDX(reverse, i, size)]);
 	}
-	GCHelper::safepoint();
-	GCHelper::detach();
+	GCHelper::safepoint(id);
+	GCHelper::detach(id);
 }
 
 void mapReader(AbstractMapProvider &map, void **keys, void **values, std::atomic<void*> *result, size_t size, bool reverse=false) {
-	GCHelper::attach();
+	map::GC::EpochType id = GCHelper::attach();
 	for (int i = 0; i < size; i++) {
 		void *res = map.get(keys[IDX(reverse, i, size)]);
 		result[IDX(reverse, i, size)].store(res, std::memory_order_relaxed);
 	}
-	GCHelper::safepoint();
-	GCHelper::detach();
+	GCHelper::safepoint(id);
+	GCHelper::detach(id);
 }
 
 void mapRemover(AbstractMapProvider &map, void **keys, void **values, std::atomic<void*> *result, size_t size, bool reverse=false) {
-	GCHelper::attach();
+	map::GC::EpochType id = GCHelper::attach();
 	for (int i = 0; i < size; i++) {
 		void *res = map.remove(keys[IDX(reverse, i, size)]);
 		result[IDX(reverse, i, size)].store(res, std::memory_order_relaxed);
 	}
-	GCHelper::safepoint();
-	GCHelper::detach();
+	GCHelper::safepoint(id);
+	GCHelper::detach(id);
 }
 
 template <bool overlap=false, bool altDirection=false>
@@ -182,7 +182,7 @@ TEST(LockFreeHashMapBasicConcurrentChecks) {
 
 	void *res;
 
-	GCHelper::attach();
+	map::GC::EpochType id = GCHelper::attach();
 
 	mapReader(map, keys, values, results, 1);
 	CHECK_ARRAY_EQUAL(nullarr, results, 1);
@@ -197,7 +197,7 @@ TEST(LockFreeHashMapBasicConcurrentChecks) {
 	CHECK_EQUAL((void*)NULL, res);
 	remover.join();
 
-	GCHelper::detach();
+	GCHelper::detach(id);
 
 	CHECK_EQUAL(0, map.unsafeUsed());
 #ifdef DEBUG_MAP_GC
@@ -319,7 +319,7 @@ TEST(LockFreeHashMapConcurrentMixedLoad) {
 			bool *conditions = new bool[bSize]();
 			int conditionsMet = 0;
 
-			GCHelper::attach();
+			map::GC::EpochType id = GCHelper::attach();
 			
 			while (conditionsMet < bSize) {
 				for (int i = 0; i < jobSize; ++i) {
@@ -334,11 +334,11 @@ TEST(LockFreeHashMapConcurrentMixedLoad) {
 						if (conditions[i + jobSize]) conditionsMet++;
 					}
 				}
-				GCHelper::safepoint();
+				GCHelper::safepoint(id);
 			}
 			delete[] conditions;
 
-			GCHelper::detach();
+			GCHelper::detach(id);
 
 			writer.join();
 			remover.join();

--- a/src/test/cpp/test_maps.cpp
+++ b/src/test/cpp/test_maps.cpp
@@ -9,7 +9,13 @@
 
 using namespace map;
 
-typedef ConcurrentMapProvider<PointerHasher<int>, false> TestLockFreeMap;
+struct TestHasher {
+	static int64_t hash(void *p) {
+  		return (int64_t)p / sizeof(int);
+	}
+};
+
+typedef ConcurrentMapProvider<TestHasher, false> TestLockFreeMap;
 typedef void (*MapFunction)(AbstractMapProvider &, void **, void **, std::atomic<void*> *, size_t, bool);
 
 void **allocateTestBuffer(size_t size) {

